### PR TITLE
ON-39558 # Added `requestBodyHeader` and returned appropriate data fr…

### PR DESCRIPTION
**Pull request by Zaxist**
…om endpoint

### CODE REVIEW

File: src/uploadToS3.ts [hunk 1]
```diff
@@ -3,36 +3,45 @@ import { Upload } from '@aws-sdk/lib-storage'
...
+  response?: T & BaseResponse
 
   async handle(request: HttpRequest, options?: HttpHandlerOptions) {
     const token = await this.getIdToken()
     if (token) {
-      request.headers['x-ob-authorization'] = 'Bearer ' + token
+      request.headers['x-oneblink-authorization'] = 'Bearer ' + token
+    }
+    if (this.requestBodyHeader) {
+      request.headers['x-oneblink-request-body-header'] = JSON.stringify(
```
Comment (mymattcarroll) [inline src/uploadToS3.ts:42]:
```suggestion
      request.headers['x-oneblink-request-body'] = JSON.stringify(
```

File: src/uploadToS3.ts [hunk 2]
```diff
@@ -54,23 +63,28 @@ const endpointSuffix = '/storage'
 export interface UploadToS3Props {
   key: string
   body: PutObjectCommandInput['Body']
-  tags: SubmissionEventTypes.S3SubmissionTags
+  requestBodyHeader?: RequestBodyHeader
+  tags: Record<string, string | undefined>
```
Comment (mymattcarroll) [inline src/uploadToS3.ts:67]:
```suggestion
  tags?: Record<string, string>
```

File: src/types.ts [hunk 3]
```diff
@@ -4,5 +4,5 @@ export type ProgressListener = (progress: ProgressListenerEvent) => void
 export interface OneBlinkUploaderProps {
   apiOrigin: string
   region: string
-  getIdToken: () => Promise<string>
+  getIdToken: () => Promise<string | void>
```
Comment (mymattcarroll) [inline src/types.ts:7]: Why `void` instead of `undefined`?

Comment (divporter) [reply src/types.ts:7]: + 1

File: src/index.ts [hunk 4]
```diff
@@ -20,10 +20,26 @@ class OneBlinkUploader {
...
+      }[]
+    }
+  } & Pick<UploadToS3Props, 'onProgress' | 'abortSignal'>) {
     console.log('Uploading submission...')
 
-    return uploadToS3({
+    return uploadToS3<{
+      uploadedAt: string
+      submissionId: string
+    }>({
```
Comment (mymattcarroll) [inline src/index.ts:42]:
```suggestion
    return uploadToS3<{
      submissionTimestamp: string
      submissionId: string
      pdfAccessToken?: string
      preventPayment: boolean
    }>({
```

File: src/index.ts [hunk 5]
```diff
@@ -20,10 +20,26 @@ class OneBlinkUploader {
...
+    tags: SubmissionEventTypes.S3SubmissionTags
+    requestBodyHeader: {
+      formsAppId?: number
+      externalId?: string
+      previousFormSubmissionId?: string
+      jobId?: string
+      taskId?: string
+      taskActionId?: string
+      taskGroupInstanceId?: string
+      recaptchas: {
```
Comment (mymattcarroll) [inline src/index.ts:32]:
```suggestion
      recaptchas?: {
```

File: src/index.ts [hunk 6]
```diff
@@ -20,10 +20,26 @@ class OneBlinkUploader {
   }: {
     body: SubmissionTypes.NewS3SubmissionData
     formId: number
-  } & Pick<UploadToS3Props, 'onProgress' | 'abortSignal' | 'tags'>) {
+    tags: SubmissionEventTypes.S3SubmissionTags
+    requestBodyHeader: {
+      formsAppId?: number
```
Comment (mymattcarroll) [inline src/index.ts:25]:
```suggestion
      formsAppId: number
```

File: src/uploadToS3.ts [hunk 7]
```diff
@@ -133,7 +149,12 @@ async function uploadToS3({
...
     await managedUpload.done()
 
-    console.log('Response', requestHandler.response)
+    if (!requestHandler.response) {
+      throw new Error(
+        'No response from server. Something went wrong in the OneBlink/uploads SDK.',
+      )
+    }
+    return requestHandler.response
   } catch (err) {
```
Comment (mymattcarroll) [inline src/uploadToS3.ts:158]: This catch can be removed, we want the abort signal to throw an error as well.

**Pull request merged**